qutip/tests/core/test_qobjevo.py

```diff
@@ -121,376 +149,234 @@ def _assert_qobjevo_different(obj1, obj2):
     assert any(obj1(t) != obj2(t) for t in np.random.rand(10) * .9 + 0.05)
 
 
-def _add(a, b):
-    return a + b
-
-def _sub(a, b):
-    return a - b
-
-def _mul(a, b):
-    return a * b
-
 def _div(a, b):
     return a / b
 
-def _matmul(a, b):
-    return a @ b
-
-def _tensor(a, b):
-    return a & b
 
-def _conj(a):
-    return a.conj()
-
-def _dag(a):
-    return a.dag()
-
-def _trans(a):
-    return a.trans()
-
-def _neg(a):
-    return -a
+class TestQobjevo:
+    def test_call(self, pseudo_qevo, coeff_type):
+        # test creation of QobjEvo and call
+        qevo = QobjEvo(*getattr(pseudo_qevo, coeff_type)())
+        assert isinstance(qevo(0), Qobj)
+        assert qevo.isoper
+        assert not qevo.const
+        assert not qevo.issuper
+        _assert_qobjevo_equivalent(pseudo_qevo, qevo)
+
+    @pytest.mark.parametrize('coeff_type',
+                             ['func_coeff', 'string',
+                              'spline', 'array', 'logarray'])
+    def test_product_coeff(self, pseudo_qevo, coeff_type):
+        # test creation of QobjEvo with Qobj * Coefficient
+        # Skip pure func: QobjEvo(f(t, args) -> Qobj)
+        base = getattr(pseudo_qevo, coeff_type)()
+        cte, [qobj, coeff] = base[0]
+        args = base[1] if len(base) >= 2 else {}
+        tlist = base[2] if len(base) >= 3 else None
+        coeff = coefficient(coeff, args=args, tlist=tlist)
+        created = cte + qobj * coeff
+        _assert_qobjevo_equivalent(pseudo_qevo, created)
+
+    def test_copy(self, all_qevo):
+        qevo = QobjEvo(*all_qevo)
+        copy = qevo.copy()
+        _assert_qobjevo_equivalent(copy, qevo)
+        assert copy is not qevo
+
+    @pytest.mark.parametrize('bin_op', [
+        pytest.param(lambda a, b: a + b, id="add"),
+        pytest.param(lambda a, b: a - b, id="sub"),
+        pytest.param(lambda a, b: a * b, id="mul"),
+        pytest.param(lambda a, b: a @ b, id="matmul"),
+        pytest.param(lambda a, b: a & b, id="tensor"),
+    ])
```

Comment:
If you find yourself reusing a parametrisation a lot, you can encode it up top in a fixture, e.g.
```python
@pytest.fixture(params=[
    pytest.param(lambda a, b: a + b, id="add"),
    ....
])
def bin_op(request):
    return request.param
```

Reply:
I also see now that you did this elsewhere in the file.